src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -604,40 +610,14 @@ public unsafe static string Join<T>(string separator, IEnumerable<T> values)
         }
 
         [ComVisible(false)]
-        public static string Join(string separator, IEnumerable<string> values)
+        public unsafe static string Join(string separator, IEnumerable<string> values)
         {
-            if (values == null)
-            {
-                throw new ArgumentNullException(nameof(values));
-            }
-
-            using (IEnumerator<string> en = values.GetEnumerator())
+            if (string.IsNullOrEmpty(separator))
+                return string.Concat(values);
+            fixed (char* pSeparator = &separator.m_firstChar)
             {
-                if (!en.MoveNext())
-                {
-                    return string.Empty;
-                }
-
-                string firstValue = en.Current;
-
-                if (!en.MoveNext())
-                {
-                    // Only one value available
-                    return firstValue ?? string.Empty;
-                }
-
-                // Null separator and values are handled by the StringBuilder
-                StringBuilder result = StringBuilderCache.Acquire();
-                result.Append(firstValue);
-
-                do
-                {
-                    result.Append(separator);
-                    result.Append(en.Current);
-                }
-                while (en.MoveNext());
-
-                return StringBuilderCache.GetStringAndRelease(result);
+                // Defer argument validation to the internal function
+                return JoinCore(pSeparator, separator.Length, values);
```

Comment:
It can be a little slower on calling `ToString()` methods that does nothing for strings. So I think it does not downgrade performance but remove copy paste code from source.